### PR TITLE
[RFC DRAFT WIP] globby@10 update (with fast-glob implementation)

### PR DESCRIPTION
**Pull request by brodybits**
- use globby@10 (globby@10.0.1) which has fast-glob implementation
- some minor test snapshot updates in tests_integration since fast-glob does not always keep the same file order as node-glob

I am raising this PR for the sake of testing and discussion related to PR #6639 (globby@7 update).

This seems to work from git on macOS and Linux, *mostly* working from git on Windows.

P.S. If we do this update, there should be no need to revert PR #6333 as discussed in #6719 / #6721

TODO items:

- [ ] resolve performance issue with finding plugins
- [ ] consider updating `CHANGELOG.unreleased.md` with information about switch to fast-glob implementation
- [ ] consider updating doc that fast-glob is only used in the new release ref: #6333 /#6719 / #6721

/cc @fisker @lydell @evilebottnawi

- ~~I’ve added tests to confirm my change works.~~
- ~~(If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)~~
- [ ] (If the change is user-facing) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
>  issue with absolute path integration test on Windows

add 

```js
patterns = patterns.map(path => path.replace(/\\/g, '/'))
```

here ,

https://github.com/prettier/prettier/blob/6d4d6a759984c40ad2648fa17ad709c7a9830efb/src/cli/util.js#L403

or follow this https://github.com/mrmlnc/fast-glob#how-to-write-patterns-on-windows

**Comment by fisker** [comment]
Another thing need attention https://dev.azure.com/prettier/prettier/_build/results?buildId=2385

this is taking tooooo long, the reason why it's too slow is because of 

https://github.com/prettier/prettier/blob/3805034bdfe45722e31a1abf019d4368dfb8ae11/src/common/load-plugins.js#L97-L104

but I'm not sure about the logic, do we find plugins in `node_modules` in any deepth or just root

**Comment by fisker** [comment]
about `regeneratorRuntime`, I think we can update babel settings, like this

https://github.com/fisker/promisify-file-reader/blob/master/babel.config.js#L9-L16 

just exclude `transform-regenerator` and add `babel-plugin-transform-async-to-promises`

**Comment by brodybits** [comment]
> add
> 
> ```js
> patterns = patterns.map(path => path.replace(/\\/g, '/'))
> ```

Thanks @fisker, seems to do the trick (with a couple Prettier lint fixes applied). I had to apply a few more updates to get some consistent results on Windows and Linux:

- revert the removal of `nodir: true` option
- add sorting of the file paths from globby (ordering did not seem to be consistent on Windows & Linux)
- revert test snapshot updates that are no longer needed

We will now see if the build results on Windows, Linux, and macOS are really consistent or not.

About the other points: unfortunately I am not so familiar with these. @fisker I wouldn't mind if you want to take this update over.

I am now starting to wonder how much we really want to switch to the fast-glob implementation, given that it is still known to be slower, seems to suffer from performance issues in certain cases, and needs special support for regeneratorRuntime. I wonder if we should consider switching to async functions (or maybe just use promises) to keep up with the newer library updates?

**Comment by fisker** [comment]
@brodybits https://github.com/brodybits/prettier/pull/9

**Comment by fisker** [comment]
> revert the removal of nodir: true option

there is no `nodir` option in v10 , same option is `onlyFiles`, and this default `true`  https://github.com/mrmlnc/fast-glob#compatible-with-node-glob

**Comment by brodybits** [comment]
> @brodybits [brodybits#9](https://github.com/brodybits/prettier/pull/9)

Thanks @fisker for the updates, I pushed these to my branch almost an hour ago.

I am finding it kinda strange that I get a green build on this PR, while I get a few tests errors in both my local workarea (on Linux) and brodybits/prettier#9. I will probably need some time to investigate these, any ideas would be much helpful.

Here are the next steps I am planning to take:

- pull updates from master
- rebase to cleanup some of the history

**Comment by fisker** [comment]
@brodybits merge master into this branch will start the ci test  runing , I have some issue with my laptop cant work on this now. you can continue work on this

File: src/common/load-plugins.js [hunk 1]
```diff
@@ -96,13 +96,15 @@ function loadPlugins(plugins, pluginSearchDirs) {
...
+      "./@prettier/plugin-*/package.json"
     ],
-    { cwd: nodeModulesDir }
+    { cwd: nodeModulesDir, expandDirectories: false }
   );
-  return pluginPackageJsonPaths.map(path.dirname);
+  return pluginPackageJsonPaths
+    .map(packageJsonFile => path.dirname(packageJsonFile.slice(2)))
+    .sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
 }
```
Comment (fisker) [inline src/common/load-plugins.js]: something is not right here, you can work on this, restore this will pass all test,  but very slow

Comment (fisker) [reply src/common/load-plugins.js]: Its late in China, going to sleep now, will back online in 9 hours, I  was trying to make this fast. maybe the slice is not right just need check the glob result. lets see how fast it will be

**Comment by brodybits** [comment]
@fisker I have a couple more questions about the updates from brodybits#9 and using fast-glob in general:

- What is the effect on the build for the online playground (browser build)? Any pointers to help me understand this part?
- Is there anything we can revert or remove once Node.js 6 support is completely gone?

**Comment by fisker** [comment]
> What is the effect on the build for the online playground (browser build)? Any pointers to help me understand this part?

the build scripts changes effects two thing

1) new package use util.promisify that node4 dont has, so I add pify to replace it

2) there is async function, babel transform into generater by default, I use a plugin to transform them into promise

> Is there anything we can revert or remove once Node.js 6 support is completely gone?

just pify thing

File: src/common/load-plugins.js (hunk 1 above)
Comment (brodybits) [reply src/common/load-plugins.js]: I reverted these changes and updated `findPluginsInNodeModules` to work more like a pipeline. This should make it easier for us to add and update the post-processing steps as needed. Unfortunately I am still not understand this well enough to think of anything else useful right now.

**Comment by brodybits** [comment]
I made some updates as discussed in <https://github.com/prettier/prettier/pull/6748#discussion_r340862955> and did a clean rebase. The build should be green now.

Thanks @fisker for answering my questions.

At this point I am not so enthusiastic about the new globby update due to the known performance disadvantages. I wish they offered an option to go back to node-glob which would be a little more deterministic (no need to use sort to get deterministic output) and not suffer from the performance drawbacks.

@fisker if you agree with this, then maybe one of us could raise an issue on globby?

I am now thinking we should consider using globby@7 for now. I would be happy to raise a new PR to do this, if people would be interested.

**Comment by fisker** [comment]
Dont give up so quickly I 'll try something else today

**Comment by fisker** [comment]
@brodybits can you pull https://github.com/fisker/prettier/tree/globby-10-update-with-fast-glob , I prefer keep the commit history, we always do `squash and merge`, no need to keep this clean

**Comment by fisker** [comment]
We need review logic from https://github.com/prettier/prettier/pull/4192 

The reason I change searching `package.json` to searching `directory` is `fast-glob` is really show without `deep` in a big dir like `node_modules`.

We'll discuss this later

**Comment by fisker** [comment]
related issues and PRs

- https://github.com/prettier/prettier/issues/6085 should be fixed
- https://github.com/prettier/prettier/pull/6285 need changing `prettier --write "**"` to  `prettier --write .` 
- see also https://github.com/prettier/prettier/pull/6285
- maybe this https://github.com/prettier/prettier/issues/4989 ?

**Comment by lydell** [comment]
Maybe we should leave globby at the current version until we improve the CLI in 2.0 or 3.0 or something.

**Comment by fisker** [comment]
@lydell why? update `globby` can benefits file globbing for users, current problem only effect plugin searching logic, we can stil use `node-glob` for this part, or even no glob at all, just need a better solution. let's discuss until my newly code get pulled here

**Comment by brodybits** [comment]
Thanks @fisker for all of your work on this.

At this point I am with @lydell to hold off on this update for now. I am really not so thrilled by the number of workarounds we have to do to deal with fast-glob. I think we should just require and use packages from npm, more complex I/O operations should be moved into reusable npm packages.

I would like to take a look through the updates, which seem to be in fisker#120, would like to work on some other priorities for now.

I would personally favor globby@7 update for now, in order to fall less far behind.

@fisker please feel free to open another new PR if you want to continue with this proposal. I would be happy to help review it if you like.

@lydell I have an off-topic comment with a question coming up.

**Comment by brodybits** [comment]
Off-topic for @lydell:

I really think we need to start a new major release asap to do the following:

- completely drop support for old Node.js versions
- consider "breaking" dependency updates
- consider any other breaking changes that should be done with very limited controversy

Relying on old dependency versions makes the update increasingly more difficult over time. Supporting old Node.js versions makes this even worse. This seems to be a major diversion of time and attention that could be used for improvements and fixes.

It is not clear to me if and where this is being discussed. I would be happy to raise a new issue, if needed. Pointers would be much appreciated.

/cc @prettier/core

**Comment by lydell** [comment]
@brodybits I completely agree with you. That’s exactly what we should be doing after 1.19. Ideally we’d release 2.0 early January 2020 with support for Node.js 10+.

**Comment by brodybits** [comment]
> Ideally we’d release 2.0 early January 2020 with support for Node.js 10+.

Is this discussed in one place somewhere?

Is there anything special that an outside contributor such as @fisker or I should do to propose breaking or potentially breaking changes for 2.0?

**Comment by lydell** [comment]
> Is this discussed in one place somewhere?

Unfortunately not. I’d suggest opening an issue like #3503 with suggested breaking changes to start a discussion.

**Comment by fisker** [comment]
@lydell @brodybits

> like it’s very difficult to upgrade

its not , I already did all the research and know how to make it work

> that the newer globby versions are slower

indeed it is slower in certain case like our plugin searching , but most cases fast-glob is faster

I will post a snippet to prove this, when i got on laptop

for plugin searching , I did make a benchmark to test  but forgot to push to github. as you can see on https://github.com/fisker/prettier/pull/120 ci is not slow anymore, but maybe best way is not touch this part, I am planning to use node-glob for this part.

**Comment by fisker** [comment]
for speed test, use your prettier codebase to test scripts bellow

```bash
node -e "console.time();require('fast-glob').sync('**/*.js');console.timeEnd();"
# default: 852.121ms

node -e "console.time();require('glob').sync('**/*.js');console.timeEnd();"
# default: 4233.523ms
```

for plugin search, `fast-glob` is slower, but this function can refactor

```bash
node -e "console.time();require('glob').sync('{*,@*/*}/package.json', {cwd: 'node_modules'});console.timeEnd();"
# default: 96.961ms

node -e "console.time();require('fast-glob').sync('{*,@*/*}/package.json', {cwd: 'node_modules'});console.timeEnd();"
# default: 261.463ms
```

results are on my laptop, windows 10 and node 10

**Comment by lydell** [comment]
Cool, thanks. You’re testing `glob` and `fast-glob` directly and not `globby`, though, but hopefully using `globby` directly gives the same results.

**Comment by fisker** [comment]
```bash
yarn add globby7@npm:globby@7 globby10@npm:globby@10
```

then

```text
E:\project\@forks\prettier>node -e "console.time();require('globby7').sync('**/*.js');console.timeEnd();"
default: 5568.626ms

E:\project\@forks\prettier>node -e "console.time();require('globby10').sync('**/*.js');console.timeEnd();"
default: 923.822ms

E:\project\@forks\prettier>node -e "console.time();require('globby7').sync('**/*.{css,html,js,json,less,md,scss,ts,vue,yaml,yml}');console.timeEnd();"
default: 10517.388ms

E:\project\@forks\prettier>node -e "console.time();require('globby10').sync('**/*.{css,html,js,json,less,md,scss,ts,vue,yaml,yml}');console.timeEnd();"
default: 997.479ms

E:\project\@forks\prettier>node -e "console.time();require('globby7').sync('{*,@*/*}/package.json', {cwd: 'node_modules'});console.timeEnd();"
default: 141.681ms

E:\project\@forks\prettier>node -e "console.time();require('globby10').sync('{*,@*/*}/package.json', {cwd: 'node_modules'});console.timeEnd();"
default: 311.261ms
```

@lydell

**Comment by lydell** [comment]
Nice! Seems like newer globby is usually a lot faster then.

**Comment by mrmlnc** [comment]
JFYI

> I'm a `fast-glob` maintainer.

The `fast-glob` package in most cases, faster than competitors. But he have a some performance problems for complex patterns (https://github.com/mrmlnc/fast-glob/issues/156) like `{*,@*/*}` or https://github.com/mrmlnc/fast-glob/issues/156#issuecomment-485327918. And, yeap, we will work on it.

**Comment by brodybits** [comment]
Closing in favor of #6776 by @fisker.

@mrmlnc let's move your comment <https://github.com/prettier/prettier/pull/6748#issuecomment-551043561> to PR #6776.

@fisker I find all of your old changes to be really hard to follow, would favor it if you would try to clean them up.

**Comment by fisker** [comment]
@brodybits I won't clean it, I  can  check what I've tryied
